services/src/main/java/org/keycloak/jose/jws/DefaultTokenManager.java

```diff
@@ -142,4 +155,96 @@ private String getSignatureAlgorithm(String clientAttribute) {
         return DEFAULT_ALGORITHM_NAME;
     }
 
+    @Override
+    public String encodeAndEncrypt(Token token) {
+        String encodedToken = encode(token);
+        if (isTokenEncyptRequired(token.getCategory())) {
+            encodedToken = getEncryptedToken(token.getCategory(), encodedToken);
+        }
+        return encodedToken;
+    }
+
+    private boolean isTokenEncyptRequired(TokenCategory category) {
+        if (cekManagementAlgorithm(category) == null) return false;
+        if (encryptAlgorithm(category) == null) return false;
+        return true;
+    }
+
+    private String getEncryptedToken(TokenCategory category, String encodedToken) {
+        String encryptedToken = null;
+
+        String algAlgorithm = cekManagementAlgorithm(category);
+        String encAlgorithm = encryptAlgorithm(category);
+
+        CekManagementProvider cekManagementProvider = session.getProvider(CekManagementProvider.class, algAlgorithm);
+        JWEAlgorithmProvider jweAlgorithmProvider = cekManagementProvider.jweAlgorithmProvider();
+
+        ContentEncryptionProvider contentEncryptionProvider = session.getProvider(ContentEncryptionProvider.class, encAlgorithm);
+        JWEEncryptionProvider jweEncryptionProvider = contentEncryptionProvider.jweEncryptionProvider();
+
+        ClientModel client = session.getContext().getClient();
+
+        KeyWrapper keyWrapper = getEncryptionKekWrapper(algAlgorithm, client);
+        if (keyWrapper == null) {
+            throw new RuntimeException("can not get encryption KEK");
+        }
+        Key encryptionKek = keyWrapper.getPublicKey();
+        String encryptionKekId = keyWrapper.getKid();
+        try {
+            encryptedToken = TokenUtil.jweKeyEncryptionEncode(encryptionKek, encodedToken.getBytes("UTF-8"), algAlgorithm, encAlgorithm, encryptionKekId, jweAlgorithmProvider, jweEncryptionProvider);
+        } catch (JWEException | UnsupportedEncodingException e) {
+            throw new RuntimeException(e);
+        }
+        return encryptedToken;
+    }
+
+    private KeyWrapper getEncryptionKekWrapper(String algAlgorithm, ClientModel client) {
+        if (algAlgorithm == null) return null;
+        switch (algAlgorithm) {
```

Comment:
Yes, I should not include algorithm specific codes in this `DefaultTokenManager`. I will clean out them.

Reply:
Thanks

Reply:
What do you think about the following codes?

```java
    private KeyWrapper getEncryptionKekWrapper(String algAlgorithm, ClientModel client) {
        if (algAlgorithm == null) return null;
        List<String> providerIds = new LinkedList<>(session.listProviderIds(CekManagementProvider.class));
        if (providerIds.contains(algAlgorithm)) return PublicKeyStorageManager.getClientPublicKeyWrapper(session, client, JWK.Use.ENCRYPTION);
        return null;
    }
```

Reply:
I am sorry I did not spot this before, but now when thinking about it and looking at it one more, I can see some potential issues with this approach. IMO the possible issue is, that algorithm is not passed to the PublicKeyStorageManager. There can be scenario when the "jwk" set returned from the client "jwks_url" contains more keys for the KeyUse.ENC for example. In that case, it would be a risk that bad key would be returned for different algorithm that you want. Filtering algorithms this way you did may not be sufficient as there can be keys and all of them could be of the algorithms, which are actually supported by Keycloak. There can be even more keys with the usage of KeyUse.ENC and for the same algorithm. The combination (keyUse, algorithm) is not guaranteed to be unique within JWK Set. Only guaranteed unique thing is "kid" AFAIK.

I wonder if we can remove method "getEncryptionKekWrapper(algAlgorithm, client)" on the DefaultTokenManager class at all. In the method "getEncryptedToken(TokenCategory category, String encodedToken)" there can be a possibility to first retrieve "kid" from the encoded token? Maybe some more changes will be needed in the class "JWE" to achieve this, but I guess this may be possible? I am seeing that retrieving "kid" first is done for example in the method "decode(String token, Class<T> clazz)", so I hope this may be possible in our case as well. Then we can lookup PublicKeyStorageManager.getClientPublicKeyWrapper by "kid". This will actually mean that PR can be simplified if I look correctly as there won't be a need to change anything in the class PublicKeyStorageManager.getClientPublicKeyWrapper and all the related classes used from that. WDYT?

Reply:
Yes, it is better to use "kid" instead of combination of "alg" and "use" in order to specify the KEK uniquely.

To do so, we need to know "kid" of KEK before encrypting the ID token.
The way hit on me is that keycloak loads Client's public keys from "jwks_url" in advance and an administrator select one of them as KEK.

The potential problem of this way is that Client's public keys loaded by keycloak does not sync with public keys Client actually manages. (e.g. Client might delete the key used for KEK marked in keycloak, and keycloak doen not know about it.)

To get around this problem, keycloak loads Client's public keys from "jwks_url"  just before encrypting the ID token every time.

However, this way needs to determine which Client's public key is used as KEK when encrypting the ID token. An administrator can not intervene manually at that time.

The next best way instead of using "kid" seems to be that (using "alg" and "enc" as filtering key) :
* get client's public keys from "jwks_url" when just encrypting the ID token
* filter out of them with keycloak specified "alg" and "use" (e.g. "alg"="RSA1_5", "use"="enc")
* if multiple public keys are selected, first one of them is used for KEK

What do you think about it?

When verifying a token in JWS, or decrypting a token in JWE, it is easy to determine "kid" because it is included in the token's JOSE header.

However, signing a token in JWS, or encrypting a token in JWE, keycloak itself needs to determine which key are used for signing or encyrpting KEK. It is easy to do so when keycloak itself manage keys used for these purpose.  That is the case for signing a token in JWS. It seems not easy to do so when keycloak itself does not manage keys used for these purpose. That is the case for encrypting a token in JWE.

Reply:
I see and agree that using "kid" is not very ideal though. Will be good to use the "alg" + "enc" combination as you proposed.

Reply:
I agree with your idea saying that method's name (`getFirstPublicKey`) explicitly tell its behavior (if multiple client public key are found, first one is retrieved).

I think PublicKeyStorageProvider.getFirstPublicKey(modelKey, algorithm, loader) seems better.
the reasons are as follows :

* JWK's "alg" implicitly determines "use" (e.g. "alg" = "RSA-OAEP" implies "use"="enc", "alg" = "RS256" implies "use"="sig") that means specifying usage of key in .getFirstPublicKey() seems redundant.

* Looking into `PublicKeyStorageManager.getClientPublicKeyWrapper(KeycloakSession session, ClientModel client, JWK.Use keyUse)` and `ClientPublicKeyLoader`, the loaded client public keys have already filtered out by key usage (if key use = "enc", only client public keys with "use" = "enc" are loaded). Therefore, it seems that no need to considering "use" when calling PublicKeyStorageProvider.getFirstPublicKey().

```java
    // Algorithm need to be specified
    public static KeyWrapper getClientPublicKeyWrapper(KeycloakSession session, ClientModel client, JWK.Use keyUse, String algAlgorithm) {
        PublicKeyStorageProvider keyStorage = session.getProvider(PublicKeyStorageProvider.class);
        String modelKey = PublicKeyStorageUtils.getClientModelCacheKey(client.getRealm().getId(), client.getId(), keyUse);
        // here client's pubic keys are loaded and filtered out by key usage
        ClientPublicKeyLoader loader = new ClientPublicKeyLoader(session, client, keyUse);
        // here only need to consider algorithm
        return keyStorage.getFirstPublicKey(modelKey, algAlgorithm, loader);
    }
```

What do you think about it?

Reply:
@tnorimat That works for me. Thanks!